core/src/main/scala/zio/config/ReadError.scala

```diff
@@ -9,6 +10,24 @@ sealed trait ReadError[A] extends Exception { self =>
     case ReadError.AndErrors(list)                => s"AndErrors(${list.map(_.toString)}"
   }
 
+  def atKey(key: A): ReadError[A] =
+    self match {
+      case ReadError.MissingValue(path)             => ReadError.MissingValue(path :+ Step.Key(key))
+      case ReadError.FormatError(path, message)     => ReadError.FormatError(path :+ Step.Key(key), message)
+      case ReadError.ConversionError(path, message) => ReadError.ConversionError(path :+ Step.Key(key), message)
+      case ReadError.OrErrors(list)                 => ReadError.OrErrors(list.map(_.atKey(key)))
+      case ReadError.AndErrors(list)                => ReadError.AndErrors(list.map(_.atKey(key)))
+    }
+
+  def atIndex(index: Int): ReadError[A] =
+    self match {
+      case ReadError.MissingValue(path)             => ReadError.MissingValue(path :+ Step.Index(index))
+      case ReadError.FormatError(path, message)     => ReadError.FormatError(path :+ Step.Index(index), message)
+      case ReadError.ConversionError(path, message) => ReadError.ConversionError(path :+ Step.Index(index), message)
+      case ReadError.OrErrors(list)                 => ReadError.OrErrors(list.map(_.atIndex(index)))
+      case ReadError.AndErrors(list)                => ReadError.AndErrors(list.map(_.atIndex(index)))
+    }
```

Comment:
There is a lot more to be done in this side to avoid redundant errors, and proper indexing.
A part of it is done in this PR. We will continue the rest in another PR (as this one is majorly map implementation and cleaning up write-to-hocon functionality)